h2o-core/src/main/java/water/parser/ParseSetup.java

```diff
@@ -538,9 +538,11 @@ private int unifyColumnCount(int cntA, int cntB, ParseSetup mergedSetup, String
       if (namesA == null) return namesB;
       else if (namesB == null) return namesA;
       else {
+        if (namesA.length != namesB.length) {
+          throw new ParseDataset.H2OParseException("Column length does not match between files.");
+        }
         for (int i = 0; i < namesA.length; i++) {
-          if (i > namesB.length || !namesA[i].equals(namesB[i])) {
-            // TODO improvement: if files match except for blanks, merge?
```

Comment:
Merging seems like a bad idea in some situations. One example.

File A (with headers):

```csv
"IsDepDelayed","fYear","fMonth","fDayofMonth","fDayOfWeek","UniqueCarrier","Origin","Dest","Distance"
"YES","f1987","f10","f14","f3","PS","SAN","SFO",447
"NO","f1987","f10","f18","f7","PS","SAN","SFO",447
"YES","f1987","f10","f19","f1","PS","SAN","SFO",447
```

File B (no headers, first row has missing column):
```csv
"YES","f1987","f10","f14","f3","PS","SAN","SFO"
"NO","f1987","f10","f18","f7","PS","SAN","SFO",447
"YES","f1987","f10","f19","f1","PS","SAN","SFO",447
"NO","f1987","f10","f21","f3","PS","SAN","SFO",447
```